pkg/cache/groupcache.go

```diff
@@ -0,0 +1,284 @@
+}
+
+func (c *Groupcache) Fetch(ctx context.Context, keys []string) map[string][]byte {
+	data := map[string][]byte{}
+
+	for _, k := range keys {
+		codec := galaxycache.ByteCodec{}
+
+		if err := c.galaxy.Get(ctx, k, &codec); err != nil {
+			level.Error(c.logger).Log("msg", "failed fetching data from groupcache", "err", err, "key", k)
+			continue
+		}
+
+		data[k] = codec
+	}
+
+	return data
+}
+
+func (c *Groupcache) Name() string {
+	return c.galaxy.Name()
+}
+
+func RegisterCacheStatsCollector(galaxy *galaxycache.Galaxy, reg prometheus.Registerer) {
+	gets := prometheus.NewDesc("thanos_cache_groupcache_get_requests_total", "Total number of get requests, including from peers.", nil, nil)
+	loads := prometheus.NewDesc("thanos_cache_groupcache_loads_total", "Total number of loads from backend (gets - cacheHits).", nil, nil)
+	peerLoads := prometheus.NewDesc("thanos_cache_groupcache_peer_loads_total", "Total number of loads from peers (remote load or remote cache hit).", nil, nil)
+	peerLoadErrors := prometheus.NewDesc("thanos_cache_groupcache_peer_load_errors_total", "Total number of errors from peer loads.", nil, nil)
+	backendLoads := prometheus.NewDesc("thanos_cache_groupcache_backend_loads_total", "Total number of direct backend loads.", nil, nil)
+	backendLoadErrors := prometheus.NewDesc("thanos_cache_groupcache_backend_load_errors_total", "Total number of errors on direct backend loads.", nil, nil)
+	cacheHits := prometheus.NewDesc("thanos_cache_groupcache_hits_total", "Total number of cache hits.", []string{"type"}, nil)
+
+	collector := &CacheStatsCollector{
+		galaxy:            galaxy,
+		gets:              gets,
+		loads:             loads,
+		peerLoads:         peerLoads,
+		peerLoadErrors:    peerLoadErrors,
+		backendLoads:      backendLoads,
+		backendLoadErrors: backendLoadErrors,
+		cacheHits:         cacheHits,
+	}
+	reg.MustRegister(collector)
+}
+
+func (s CacheStatsCollector) Collect(ch chan<- prometheus.Metric) {
+	stats := s.galaxy.Stats
+	ch <- prometheus.MustNewConstMetric(s.gets, prometheus.CounterValue, float64(stats.Gets))
+	ch <- prometheus.MustNewConstMetric(s.loads, prometheus.CounterValue, float64(stats.Loads))
+	ch <- prometheus.MustNewConstMetric(s.peerLoads, prometheus.CounterValue, float64(stats.PeerLoads))
+	ch <- prometheus.MustNewConstMetric(s.peerLoadErrors, prometheus.CounterValue, float64(stats.PeerLoadErrors))
+	ch <- prometheus.MustNewConstMetric(s.backendLoads, prometheus.CounterValue, float64(stats.BackendLoads))
+	ch <- prometheus.MustNewConstMetric(s.backendLoadErrors, prometheus.CounterValue, float64(stats.BackendLoadErrors))
+	ch <- prometheus.MustNewConstMetric(s.cacheHits, prometheus.CounterValue, float64(stats.MaincacheHits), galaxycache.MainCache.String())
+	ch <- prometheus.MustNewConstMetric(s.cacheHits, prometheus.CounterValue, float64(stats.HotcacheHits), galaxycache.HotCache.String())
+}
+
+func (s CacheStatsCollector) Describe(ch chan<- *prometheus.Desc) {
```

Comment:
These need to be:
```suggestion
func (s *CacheStatsCollector) Describe(ch chan<- *prometheus.Desc) {
```

Since you are `Register()`ing a pointer to `CacheStatsCollector` :P